src/storage/accessors/DataAccessor.ts

```diff
@@ -0,0 +1,81 @@
+ *  * It should always throw a NotFoundHttpError if it does not have data matching the input identifier.
+ *  * The only exception is the `getNormalizedMetadata` function where both versions of the identifier
+ *    need to be checked (both with and without slash).
+ *  * DataAccessors are responsible for generating the relevant containment triples for containers.
+ */
+export interface DataAccessor {
+  /**
+   * Should throw an UnsupportedHttpError if the DataAccessor does not support storing the given Representation.
+   * @param representation - Incoming Representation.
+   *
+   * @throws UnsupportedHttpError
+   * If it does not support the incoming data.
+   */
+  canHandle: (representation: Representation) => Promise<void>;
+
+  /**
+   * Returns a data stream stored for the given identifier.
+   * It can be assumed that the incoming identifier will always correspond to a data resource.
+   * @param identifier - Identifier for which the data is requested.
+   */
+  getData: (identifier: ResourceIdentifier) => Promise<Readable>;
+
+  /**
+   * Returns the metadata corresponding to the identifier.
+   * @param identifier - Identifier for which the metadata is requested.
+   */
+  getMetadata: (identifier: ResourceIdentifier) => Promise<RepresentationMetadata>;
+
+  /**
+   * Returns the metadata matching the identifier, ignoring the presence of a trailing slash or not.
+   * This is used to support the following part of the spec:
+   * "If two URIs differ only in the trailing slash, and the server has associated a resource with one of them,
+   *  then the other URI MUST NOT correspond to another resource."
+   * @param identifier - Identifier that needs to be checked.
+   */
+  getNormalizedMetadata: (identifier: ResourceIdentifier) => Promise<RepresentationMetadata>;
+
+  /**
+   * Writes data and metadata for a data resource.
+   * If any data and/or metadata exist for the given identifier, it should be overwritten.
+   * @param identifier - Identifier of the resource.
+   * @param data - Data to store.
+   * @param metadata - Optional metadata to store.
+   */
+  writeDataResource: (identifier: ResourceIdentifier, data: Readable, metadata?: RepresentationMetadata) =>
+  Promise<void>;
+
+  /**
+   * Writes metadata for a container.
+   * If the container does not exist yet it should be created,
+   * if it does its metadata should be overwritten, except for the containment triples.
+   * @param identifier - Identifier of the container.
+   * @param metadata - Optional metadata to store.
+   */
+  writeContainer: (identifier: ResourceIdentifier, metadata?: RepresentationMetadata) => Promise<void>;
+
+  /**
+   * Modifies the resource based on the given Patch.
+   * @param identifier - Resource to modify.
+   * @param patch - Patch specifying how the resource should be modified.
+   */
+  modifyResource: (identifier: ResourceIdentifier, patch: Patch) => Promise<void>;
```

Comment:
Was a bit surprised to see that one here.

Reply:
This is something we discussed in a meeting before the work on this was started. The problem is that `modifyResource` is completely dependend on the backend storage. E.g. a file backend can't handle patches but a SPARQL backend can send the update request to the store. There is not really a way around this (except not allowing any backend to have custom patch implementations but that's not ideal), so there needs to be a way to send the patch to the backend. The function can always be renamed to something else of course.

Reply:
> (except not allowing any backend to have custom patch implementations but that's not ideal)

Yeah, this was my suggestion as I was reading this. But perhaps a bit too harsh.
That seems like the kind of optimization for which you'd implement a store then instead of an accessor.

Reply:
To combine with the comment about `getNormalizedMetadata`. When making the SPARQL store, this would mean that we would have a `SparqlDataAccessorBasedStore` (or a name with less than 5 words) and a `SparqlDataAccessor`. The store would inherit from the `DataAccessorBasedStore`. The store would override 2 functions: `modifyResource` and `getNormalizedMetadata` (now being a protected function of the store). For both of these, SPARQL queries are required, so the store would have to call `doSparql`, which is a specific function of a `SparqlDataAccessor`. So that would mean this specific store only works with this specific kind of data accessor.

Not saying that's a bad thing, none of the actual store "behaviour" is duplicated, that's all still stored in the original store, and that is the main important part of this PR for me. Just want to clear up what the result of this is.

Reply:
1. Would it also _work_ when plugging in `SparqlDataAccessor` into a normal `DataAccessorBasedStore`? So "work" as in functionally identical, just less optimal?

2. Instead of `SparqlDataAccessorBasedStore`, how about `SparqlDataAccessor` actually implements _two_ interfaces: `DataAccessor` and something like `NormalizedDataAccessor`? Then `DataAccessorBasedStore` can check for the optional presence of the second interface, and adjust accordingly?

3. To what sense are we strongly optimizing here for SPARQL though? I.e., would the Normalized pattern also occur in other cases>

Reply:
1. Yes, there would only be a few more requests.
2. That would also work. Not sure yet what my opinion is on this compared to inheriting from the store. My first gut reaction is that this might make the store messier since for several functions it might need to check if they exist. Another potential option this made me think of is having the class extend both the store and implement the data accessor interface so all sparql related stuff is contained in the same class, so it would be its own data accessor (but still use most of the generic functions of the default store). But that would be even tighter coupling of course.
3. That I don't know and is hard to predict. This is relevant specifically for backends where doing an "access" is expensive. In the case of the file store, it's a lot cheaper to check if a file exists on the file system than the cost of sending a query to a triple store.

But I am ok with moving the `getNormalizedMetadata` to the store, it makes the data accessor even cleaner. But if we ever want to optimize in the future it is something to keep in mind. The good thing is that this doesn't make optimizations impossible (or even super hard).

Reply:
Indeed, sounds good. Derivatives of `SparqlDataAccessorBasedStore` just seem to defeat the purpose (only a little though, I still see the reuse).